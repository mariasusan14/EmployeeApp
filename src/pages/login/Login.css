
.container{
    display: flex;
    flex-direction: row;
    height: 100vh;
 box-sizing: border-box;
}

.leftcont{
    background-color: #e6f7fd;
    width: 50%;
    height: 100vh;
    display: flex;
    justify-content: center;
    align-items: center;
    justify-content: center;
    overflow: hidden;
}

.circle{
    width: 60%;
    height: 60%;
    border-radius: 50%;
    background-color: white;
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
    align-items: center;
    overflow: hidden;
    padding: 20px;
}
.rightcont{
   display: flex;
   flex-direction: column;
   justify-content: center;
   align-items: center;
   width: 50%;
   height: 100vh;
}
.login{
    position: relative;
    display: flex;
    flex-direction: column;
    align-items: left;
    width: 100%;
    
}

.rightcont >.form{
    width: 50%;
    box-shadow: none;
}

.username-transition{
    position: absolute;
    top: 20%;
    left: 10px;
    background-color: white;
    padding: 1px;
    color: #03AEEE;
    font-size: small;
}

.username-input-wrapper{
    margin-top: 20px;
   
}
.login .inputWrapper{
    width: 100%;
    margin: 0px 0px 20px 0px;
}

.errormsg{
    font-size: small;
    color: red;
    padding: 0;
    margin: 0%;
}

.checkbx{
    display: flex;
    flex-direction: row;
    align-items: flex-start;
}


.wrapper{
    display: flex;
    flex-direction: row;
    width: 100%;
}

.checkbx >.inputWrapper{
    margin: 0%;
    padding: 0%;
    max-width: fit-content;
}

.checkbx >p{
   font-size: medium;
   margin: 0;
}
.login .clear{
    color: black;
    font-weight: normal;
    width: fit-content;
    height: fit-content;
    border: 1px solid rgb(189, 186, 186);
background-color: rgb(235, 233, 233);
padding: 10px 10px;
}

