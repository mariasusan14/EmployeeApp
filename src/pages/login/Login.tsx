import "./Login.css";
import { Button } from "../../components/button/Button";
import { Input } from "../../components/input/input";
import logo from "../../assets/kv-logo.png";
import loginImg from "../../assets/kv-login.jpeg";
import { useEffect, useState, useRef } from "react";
//import { useMouse } from "../../hooks/useMouse";
import {useLocalStorage} from '../../hooks/useLocalStorage'
import { useNavigate } from "react-router-dom";
import { useLoginMutation } from "../../api-service/auth/login.api";
export const Login = () => {
  const [userName, setUsername] = useState("");
  const [error,setError]=useState("")
  const [password, setPassword] = useState("");
  const [usernameError, setUsernameError] = useState("");
  const [passwordError, setPasswordError] = useState("");
  const [showPassword, setShowPassword] = useLocalStorage("setPasswordCheckbox",false);

  const userNameRef = useRef<HTMLInputElement>(null);

  //const mouse = useMouse();
  const [login,{isLoading}]=useLoginMutation();
  const navigate = useNavigate();

  useEffect(() => {
    if (password.length < 8) {
      setPasswordError("Must be 8 characters Long");
    } else {
      setPasswordError("");
    }
    
  }, [password]); 

  useEffect(() => {
    if (userName.length > 20) {
      setUsernameError("Username must be less than 20 characters");
    } else {
      setUsernameError("");
    }
  }, [userName]);

  useEffect(() => {
    userNameRef.current?.focus();
  }, []);

  

  async function handleFormSubmit() {
login({email:userName,password:password})
.unwrap()
.then((response)=>{
  if(!response)
    throw new Error("Response not found")
localStorage.setItem("token",response.accessToken)
localStorage.setItem("isLoggedIn","true")
navigate("/employee");
})
.catch((error)=>{
  setError(error.data.error) 
})
    
  }

  function handleClear() {
    setUsername("");
  }


  return (
    <>
      <div className="container">
        <div className="leftcont">
          <div className="circle">
            <img
              src={loginImg}
              alt="kv-login"
              id="cirleimg"
              width="100%"
              height="100%"
            />
          </div>
        </div>
        <div className="rightcont">
          <form className="form">
            <div className="login">
              <img src={logo} alt="kv-logo" width="200px" height="45px"></img>
              <div className="username-input-wrapper">
                <div className="username-transition">Username</div>
                <Input
                  type="email"
                  name="username"
                  placeholder="username"
                  className="userNameInp"
                  value={userName}
                  ref={userNameRef}
                  onChange={(event) => {
                    setUsername(event.target.value);
                  }}
                  endAdornment={
                    <button
                      className="clear"
                      onClick={handleClear}
                      disabled={userName.length === 0}
                    >
                      Clear
                    </button>
                  }
                />
              </div>
               <p className="errormsg">{usernameError}</p>
              <p className="errormsg">{error}</p>
              <Input
                id="password"
                type={showPassword ? "text" : "password"}
                name="password"
                placeholder="password"
                value={password}
                onChange={(event) => {
                  setPassword(event.target.value);
                }}
                // endAdornment={
                //   <button className="clear" onClick={handleClear}>
                //     Clear
                //   </button>
                // }
              />
              <p className="errormsg">{password && passwordError}</p>
              <div className="checkbx">
                <Input
                  type="checkbox"
                  id="checkbox"
                  checked={showPassword}
                  onChange={() => {
                    setShowPassword(!showPassword);
                  }}
                />
                <p>Show Password</p>
              </div>
              <Button
                onClick={handleFormSubmit}
                type="button"
                title="Login"
                variant="primary"
                className="login-button"
                disabled={!password||!userName||isLoading}
              />
              
              {/* <div>{mouse.x}</div>
            <div>{mouse.y}</div> */}
            </div>
          </form>
        </div>
      </div>
    </>
  );
};
