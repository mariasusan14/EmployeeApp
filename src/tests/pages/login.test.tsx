import { describe, it, expect, beforeEach, vi } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import userEvent from "@testing-library/user-event";
import {Login} from "../../pages/login/Login";
// Mock react-router-dom's useNavigate
const mockNavigate = vi.fn();
vi.mock("react-router-dom", async () => {
  const actual = await vi.importActual("react-router-dom");
  return {
    ...actual,
    useNavigate: () => mockNavigate,
  };
});
// Mock the store
const mockStore = configureStore({
  reducer: {
    employee: (state = { employees: [] }) => state,
    department: (state = { departments: [] }) => state,
    employeeApi: (state = {}) => state,
  },
});
// Mock the login mutation
const mockLoginMutation = vi.fn();
vi.mock("../../api-service/auth/login.api", () => ({
  useLoginMutation: () => [
    () => ({
      unwrap: () => mockLoginMutation(),
    }),
    { isLoading: false },
  ],
}));
describe("Login Page", () => {
  const renderLogin = () => {
    return render(
      <Provider store={mockStore}>
        <Login />
      </Provider>
    );
  };
  beforeEach(() => {
    vi.clearAllMocks();
  });
  it("should match snapshot", () => {
    const { container } = renderLogin();
    expect(container).toMatchSnapshot();
  });
  it("should render login form with all required elements", () => {
    renderLogin();

    // Check if username input is present
    const usernameInput = screen.getByLabelText("username");
    expect(usernameInput).toBeInTheDocument();

    // Check if password input is present
    const passwordInput = screen.getByLabelText("password");
    expect(passwordInput).toBeInTheDocument();

    // Check if submit button is present
    const submitButton = screen.getByRole("button", { name: /login/i });
    expect(submitButton).toBeInTheDocument();
    expect(submitButton).toHaveAttribute("type", "button");
  });
  
  it("should show error when username is too long", async () => {
    renderLogin();
    const usernameInput = screen.getByLabelText("username");
    // Type a username longer than 30 characters
    await userEvent.type(usernameInput, "a".repeat(31));
    // Check if error message is displayed
    expect(
      screen.getByText("Username must be less than 20 characters")
    ).toBeInTheDocument();
  });


   it("should handle successful login", async () => {
    mockLoginMutation.mockResolvedValueOnce({ token: "fake-token" });

    renderLogin();

    // Fill in the form
    const usernameInput = screen.getByLabelText("username");
    const passwordInput = screen.getByLabelText("password");
    const submitButton = screen.getByRole("button", { name: /login/i });

    await userEvent.type(usernameInput, "test@example.com");
    await userEvent.type(passwordInput, "password123");

    // Submit the form
    await userEvent.click(submitButton);

    // Wait for the login mutation to resolve and navigation to occur
    await waitFor(() => {
      expect(mockLoginMutation).toHaveBeenCalled();
      expect(mockNavigate).toHaveBeenCalledWith("/employee");
    });
  });


});