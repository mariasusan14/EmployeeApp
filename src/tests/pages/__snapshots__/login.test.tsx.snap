// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login Page > should match snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="leftcont"
    >
      <div
        class="circle"
      >
        <img
          alt="kv-login"
          height="100%"
          id="cirleimg"
          src="/src/assets/kv-login.jpeg"
          width="100%"
        />
      </div>
    </div>
    <div
      class="rightcont"
    >
      <form
        class="form"
      >
        <div
          class="login"
        >
          <img
            alt="kv-logo"
            height="45px"
            src="/src/assets/kv-logo.png"
            width="200px"
          />
          <div
            class="username-input-wrapper"
          >
            <div
              class="username-transition"
            >
              Username
            </div>
            <div
              class="inputWrapper"
            >
              <div
                class="wrapper"
              >
                <input
                  aria-label="username"
                  class="userNameInp"
                  name="username"
                  placeholder="username"
                  required=""
                  type="text"
                  value=""
                />
                <button
                  class="clear"
                  disabled=""
                >
                  Clear
                </button>
              </div>
            </div>
          </div>
          <p
            class="errormsg"
          />
          <p
            class="errormsg"
          />
          <div
            class="inputWrapper"
          >
            <div
              class="wrapper"
            >
              <input
                aria-label="password"
                id="password"
                name="password"
                placeholder="password"
                required=""
                type="password"
                value=""
              />
            </div>
          </div>
          <div
            class="checkbx"
          >
            <div
              class="inputWrapper"
            >
              <div
                class="wrapper"
              >
                <input
                  id="checkbox"
                  required=""
                  type="checkbox"
                />
              </div>
            </div>
            <p>
              Show Password
            </p>
          </div>
          <button
            class="button button--primary login-button"
            type="button"
          >
            Login
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
